lib/player.js

```diff
@@ -1209,527 +1013,317 @@ shaka.Player = class extends shaka.util.FakeEventTarget {
-    // are not overriding the destroyed state because we will unload when we are
-    // destroying the player.
-    if (this.loadMode_ != shaka.Player.LoadMode.DESTROYED) {
-      this.loadMode_ = shaka.Player.LoadMode.NOT_LOADED;
-    }
-
-    // Run any general cleanup tasks now.  This should be here at the top, right
-    // after setting loadMode_, so that internal components still exist as they
-    // did when the cleanup tasks were registered in the array.
-    const cleanupTasks = this.cleanupOnUnload_.map((cb) => cb());
-    this.cleanupOnUnload_ = [];
-    await Promise.all(cleanupTasks);
+  async load(assetUri, startTime = null, mimeType) {
+    try {
+      // Do not allow the player to be used after |destroy| is called.
+      if (this.loadMode_ == shaka.Player.LoadMode.DESTROYED) {
+        throw this.createAbortLoadError_();
+      }
 
-    // Dispatch the unloading event.
-    this.dispatchEvent(
-        this.makeEvent_(shaka.util.FakeEvent.EventName.Unloading));
+      if (this.assetUri_) {
+        await this.unload();
+      }
 
-    // Remove everything that has to do with loading content from our payload
-    // since we are releasing everything that depended on it.
-    has.mimeType = null;
-    has.startTime = null;
-    has.uri = null;
+      this.startTime_ = startTime;
+      this.fullyLoaded_ = false;
 
-    // Release the region timeline, which is created when parsing the manifest.
-    if (this.regionTimeline_) {
-      this.regionTimeline_.release();
-      this.regionTimeline_ = null;
-    }
+      // We dispatch the loading event when someone calls |load| because we want
+      // to surface the user intent.
+      this.dispatchEvent(this.makeEvent_(
+          shaka.util.FakeEvent.EventName.Loading));
 
-    // In most cases we should have a media element. The one exception would
-    // be if there was an error and we, by chance, did not have a media element.
-    if (has.mediaElement) {
-      this.loadEventManager_.removeAll();
-    }
+      const startTimeOfLoad = Date.now() / 1000;
 
-    // Stop the variant checker timer
-    this.checkVariantsTimer_.stop();
+      // Stats are for a single playback/load session. Stats must be initialized
+      // before we allow calls to |updateStateHistory|.
+      this.stats_ = new shaka.util.Stats();
 
-    // Some observers use some playback components, shutting down the observers
-    // first ensures that they don't try to use the playback components
-    // mid-destroy.
-    if (this.playheadObservers_) {
-      this.playheadObservers_.release();
-      this.playheadObservers_ = null;
-    }
+      this.assetUri_ = assetUri;
 
-    // Stop the parser early. Since it is at the start of the pipeline, it
-    // should be start early to avoid is pushing new data downstream.
-    if (this.parser_) {
-      await this.parser_.stop();
-      this.parser_ = null;
-      this.parserFactory_ = null;
+      const shouldUseSrcEquals =
```

Comment:
I think that for simplicity we should try to obtain the mimetype before instead of each function itself trying to infer it.

First we try to obtain the mimetype of the extension, and if not we make a request. But this way we simplify other parts of the code (manifest parser for example, we could even eliminate the registration of parsers by extension) and we always make sure we have the mimetype. And if the mimetype cannot be obtained it would throw an error. what do you think?

Reply:
Yeah, that sounds like a good idea.